### ISSUE TITLE
Support @include in VS Code

### ISSUE THREAD

**Issue by babig** [christianvoigt/argdown#150]
argdown/Cli supports including source files with @include. It would be great to have this or a similar feature available in the VS Code Extension.

**Comment by christianvoigt** [comment]
Yes, I basically agree with you, but think we need to move away from the current solution to a less "hacky" import feature. 

The "@import" feature of @argdown/cli was a quick and dirty solution. It does not work with "Intellisense" in VSCode (diagnostics, code completion and so on). Before parsing the text, the "@import" strings are simply replaced by the imported files recursively, line numbers change and any code issues can not be traced back to their original files.

That's one of the reasons why I would like to implement a different concept of "importable modules" in Argdown, where you can refer to elements from other Argdown sources and those dependencies are resolved by the parser *during the parsing process*. 

Importable modules would also open up a lot of other possibilities beyond those of @import, especially in multi-user contexts. I already started prototyping this a year ago, but had to stop as it was taking too long to get this right. At the moment I don't know when I will find the time to finish this feature.

**Comment by Justin-W** [comment]
I would love to be able to create an argdown project in VSCode which supports splitting a huge argdown into multiple `.argdown` files, all of which support project-wide intellisense.

I'm imagining a wiki-like collection of statements and arguments, all of which can be cross-linked where appropriate using argdown relations, and where the entire knowledge/argument corpus can be queried, visualized, etc. Such that the project as a whole constitutes a type of "knowledge graph" of truth bearing statements, related argument reconstructions (for and against), and related sources, citations, etc.

It seems like the existing @include and partials features were an attempt to enable something like I described above, but only for the argdown CLI? But it seems like something of significant complexity, when split across multiple large argdown files, would essentially require project-wide intellisense to enable effective incremental development and maintenance of the knowledge graph/code. Especially if multiple contributors were involved.

@christianvoigt Are there any plans to continue/restart work on this type of feature?

**Comment by ggbetz** [comment]
TLDR; We're in the midst of a code-base renovation and am not planning to add new features in the near future at this moment in time.

Hi @Justin-W , thanks for using argdown and reaching out!

> I would love to be able to create an argdown project in VSCode which supports splitting a huge argdown into multiple .argdown files

That is the way to go; I'm also organizing argdown reconstruction projects this way.

> It seems like the existing @include and partials features were an attempt to enable something like I described above, but only for the argdown CLI?

Yes, intellisense / the language server doesn't seem to resolve argument and thesis workspace wide, but only file-wise. But @include works fine with the argdown-cli.

I'm using VS Code tasks / makefiles to automatize any build processes with argdown-cli in vs-code, which has so far worked great for projects like this. Would you be interested in setting up a **GitHub template repo** with scripts, makefile, task configs, folder structure, dummy argdown files, dummy source files in place?

@christianvoigt Please feel free to add!

**Comment by Justin-W** [comment]
> Yes, intellisense / the language server doesn't seem to resolve argument and thesis workspace wide, but only file-wise.
...
> I'm using VS Code tasks / makefiles to automatize any build processes with argdown-cli in vs-code, which has so far worked great for projects like this.

@ggbetz Build/compile scripts seem like a reasonable workaround for automating the graph (pre)rendering process. But how do you or other argdown experts work around the incomplete intellisense when your split files have bidirectional cross-references?

My current argdown-based project is currently a single file with >5K lines of argdown statements and argument reconstructions, >400 primary Statements, >1K Statement references, and >100 arguments. And that represents only an initial scaffolding/skeleton.

I'm not sure how I would effectively avoid typos, accidental/unintentional duplication, broken references, etc. without the use of intellisense and the "problems" pane during editing (other than by constantly using project-wide search and tab swapping, which seems like an impractical workflow from both a productivity and cognition perspective). I currently rely quite heavily on intellisense statement completion, and IDE-assisted navigation and refactoring (e.g. "Go to Definition/References", Find All References, Rename, etc.) during editing. Any tips or advice on your development/coding process while incrementally adding/editing/refining the arguments within a massive project would be appreciated.

>  Would you be interested in setting up a **GitHub template repo** with scripts, makefile, task configs, folder structure, dummy argdown files, dummy source files in place?

I would definitely be very interested in getting access to such a template repo and using it as the basis for my current and future argdown projects. I hate reinventing wheels. 😄 

Thanks!

**Comment by ggbetz** [comment]
Thanks @Justin-W for the further explanations! I see.

> My current argdown-based project is currently a single file with >5K lines of argdown statements and argument reconstructions, >400 primary Statements, >1K Statement references, and >100 arguments. And that represents only an initial scaffolding/skeleton.
> 
> I'm not sure how I would effectively avoid typos, accidental/unintentional duplication, broken references, etc. without the use of intellisense and the "problems" pane during editing (other than by constantly using project-wide search and tab swapping, which seems like an impractical workflow from both a productivity and cognition perspective).

The only way I can think of (workaround) is to split this into subdebates that have minimal connections to other subdebates. But I see that better intellisense for multiple files in folder / included files would be very good indeed.

Also, I have no template repo to share, yet. It's on my todo list, though...

**Comment by ggbetz** [comment]
An initial plan co-drafted with AI for future reference: 




### The Proposal

**Mechanism:**
1. Each Argdown file can optionally declare a `main` field in its YAML frontmatter
2. Multi-file IntelliSense context would be the **union** of:
   - Current file + all files it recursively `@include`s
   - Root file (from `main:` field) + all files it recursively `@include`s

**Example:**
```argdown
===
main: ../semmelweis_betz.argdown
title: Evidence Section
===

[Some Evidence]: This is discussed in the main debate.
```

### Strengths

**1. Declarative & Explicit**
- Users explicitly declare project structure
- No guessing about which file is "main"
- Clear intent: "this partial belongs to that root file"

**2. Leverages Existing Infrastructure**
- Frontmatter parsing already exists via `DataPlugin`
- `IncludePlugin` already handles recursive inclusion
- File path resolution logic already implemented

**3. Bidirectional Context**
- Works for both "bottom-up" (partial → main) and "top-down" (main → partials)
- Editing a partial gives you context from the main file
- Editing the main file gives you context from all partials

**4. Performance Efficient**
- Clear scope boundaries (only process 2 trees max)
- No workspace-wide scanning required
- Predictable cache invalidation (track dependencies of 2 files)

**5. Solves Common Workflows**
- **Modular debates**: Split large arguments into files, all reference main
- **Reusable components**: Common arguments/statements defined in partials
- **Team collaboration**: Different authors work on sections, all point to master

### Implementation Strategy

```typescript
interface MultiFileContext {
  currentFileTree: string[];  // Current file + includes
  mainFileTree: string[];     // Main file + includes (if main exists)
}

async function buildMultiFileContext(
  filePath: string,
  fileContent: string
): Promise<MultiFileContext> {
  
  // 1. Parse frontmatter to check for 'main' field
  const frontMatter = parseFrontMatter(fileContent);
  const mainFilePath = frontMatter?.main 
    ? path.resolve(path.dirname(filePath), frontMatter.main)
    : null;
  
  // 2. Build current file's include tree
  const currentFileTree = await buildIncludeTree(filePath, fileContent);
  
  // 3. Build main file's include tree (if exists)
  let mainFileTree: string[] = [];
  if (mainFilePath && mainFilePath !== filePath) {
    const mainContent = await readFile(mainFilePath);
    mainFileTree = await buildIncludeTree(mainFilePath, mainContent);
  }
  
  return { currentFileTree, mainFileTree };
}

async function buildIncludeTree(
  filePath: string, 
  content: string,
  visited: Set<string> = new Set()
): Promise<string[]> {
  if (visited.has(filePath)) return [];
  visited.add(filePath);
  
  const tree = [filePath];
  const includeMatches = content.matchAll(/@include\(([^)]+)\)/g);
  
  for (const match of includeMatches) {
    const includePath = path.resolve(path.dirname(filePath), match[1]);
    const includeContent = await readFile(includePath);
    const subTree = await buildIncludeTree(includePath, includeContent, visited);
    tree.push(...subTree);
  }
  
  return tree;
}
```

### Caching Strategy

```typescript
interface FileContextCache {
  // Map from file path to its context
  contexts: Map<string, MultiFileContext>;
  
  // Dependency graph: which files depend on which
  dependencies: Map<string, Set<string>>;
  
  // File modification times
  mtimes: Map<string, number>;
}

function invalidateCache(changedFile: string, cache: FileContextCache) {
  // Invalidate the changed file
  cache.contexts.delete(changedFile);
  
  // Invalidate all files that depend on it
  const dependents = cache.dependencies.get(changedFile) || new Set();
  for (const dependent of dependents) {
    cache.contexts.delete(dependent);
  }
}
```

### Real-World Usage Example

```
project/
  ├── main-debate.argdown          # Root file
  │   ===
  │   title: Semmelweis Debate
  │   ===
  │   @include(sections/hypotheses.argdown)
  │   @include(sections/evidence.argdown)
  │   @include(sections/arguments.argdown)
  │
  ├── sections/
  │   ├── hypotheses.argdown       # Partial 1
  │   │   ===
  │   │   main: ../main-debate.argdown
  │   │   ===
  │   │   [H1]: First hypothesis...
  │   │
  │   ├── evidence.argdown         # Partial 2
  │   │   ===
  │   │   main: ../main-debate.argdown
  │   │   ===
  │   │   [E1]: Evidence supports @[H1]
  │   │
  │   └── arguments.argdown        # Partial 3
  │       ===
  │       main: ../main-debate.argdown
  │       ===
  │       <A1>: @[E1] -> [H1]
```

**IntelliSense Benefits:**
- Editing `evidence.argdown`: Can autocomplete `[H1]` from hypotheses.argdown
- Editing `arguments.argdown`: Can find references to `<A1>` in main file
- Go to Definition from `@[E1]` in arguments.argdown jumps to evidence.argdown

**Comment by Morstis** [comment]
Hi everyone, 
I will start working on this issue.
@ggbetz thanks for the proposal. I have a small question regarding this:
If I understand you correct you want a tree structure where the higher nodes have accessibility to everything defined beneath. Therefore a child node that is only referenced by @include from a parent can only see its own definitions. Once the child node defines "main" it has access to all definitions of main <root> and all of the direct children of <root> (only @include nodes). So therefore if you use inside of <file1> @include <child> and inside of <child> there is a "main <root2>", <file1> should not have access to definitions of <root> or any of <root>'s children?


# Edge Case
```
project/
  ├── main-debate.argdown  # Root file has access to definitions of hypotheses + evidence, but can not access [Another]
  │   ===
  │   title: Semmelweis Debate
  │   ===
  │   @include(sections/hypotheses.argdown)
  │   @include(sections/evidence.argdown)
  |
  ├── another-debate.argdown          # Second Root
  │   ===
  │   title: Another Debate
  │   ===
  │   @include(sections/arguments.argdown)
  |     [Another]: Debate
  │
  ├── sections/
  │   ├── hypotheses.argdown       # Partial 1 Access to <E1> but not to <A1>
  │   │   ===
  │   │   main: ../main-debate.argdown
  │   │   ===
  │   │   [H1]: First hypothesis...
  │   │
  │   ├── evidence.argdown         # Needs definitions from another debate <A1>
  │   │   ===
  │   │   main: ../another-debate.argdown
  │   │   ===
  │   │   [E1]: Evidence supports @[H1]
  │   │
  │   └── arguments.argdown        # Has access to <H1> and <E1>
  │       ===
  │       main: ../main-debate.argdown
  │       ===
  │       <A1>: @[E1] -> [H1]

```

**Comment by ggbetz** [comment]
Thanks for these questions @Morstis !

> If I understand you correct you want a tree structure where the higher nodes have accessibility to everything defined beneath.

yes

> Therefore a child node that is only referenced by @include from a parent can only see its own definitions. 

yes

> Once the child node defines "main" it has access to all definitions of main ...

yes

> ... and all of the direct children of (only @include nodes).

🤔 I'd say it has also access to all definitions of direct and indirect children of the "main" file.

> So therefore if you use inside of @include and inside of there is a "main ", should not have access to definitions of or any of 's children?

🤔 ??? 

-----

Let's say every Argdown file is a node. Then `@include` defines directed edges between nodes. We require that this yields a **DAG** `G = <V,E>`. We also require that  `v` may only declare node `v°` as main file if `v` is a descendant of `v°`.

If a file `v` declares no main file, then `v` has access to all definitions defined in any descendent `u` of `v` (i.e., every file `u` such that there is a directed path from `v` to `u`). 

If a file `v` declares a main file `v°`, then `v` has access to all definitions defined in any descendent of `v°`.

Would that make sense?
